### PR TITLE
[PLT-7109 + PLT-7050] Test Run state retain + View HTML Report

### PR DESCRIPTION
**Pull request by amnambiar**
https://input-output.atlassian.net/browse/PLT-7109
https://input-output.atlassian.net/browse/PLT-7050

@coderabbitai: ignore

### Summary by CodeRabbit

- New Feature: Added a new route "/report/:uuid" to view certification results.
- New Feature: Introduced `CertificationResult` component for displaying the result of a certification process.
- Refactor: Updated Dockerfile and related scripts for improved containerization.
- Refactor: Enhanced `Certification` component with better state management, error handling, and form submission.
- Refactor: Modified `useLocalStorage` hook to include a function that removes stored value from local storage.
- Chore: Cleaned up unused files and directories, updated .gitignore and .dockerignore files.
- Chore: Updated GitHub workflows for CI/CD.
- Bug Fix: Fixed image path in `TimelineItem` component.
- Bug Fix: Improved logout logic in `ConnectWallet`, `Header`, and `UserProfile` components.

### CODE REVIEW

File: src/pages/certification/Certification.tsx [hunk 1]
```diff
@@ -204,20 +223,15 @@ const Certification = () => {
...
 
-  // while unmount of component
-  useEffect(() => {
-    return () => {
-      resetStates();
-    };
-  }, []);
 
   useEffect(() => {
     runStatus === "certifying" ? setRefetchMin(2) : setRefetchMin(5);
```
[the remark this answers is not in the document]
Comment (amnambiar) [reply src/pages/certification/Certification.tsx:245]: Ignoring this suggestion for now

File: src/pages/certification/Certification.tsx [hunk 2]
```diff
@@ -204,20 +223,15 @@ const Certification = () => {
...
   useEffect(() => {
-    if (uuid.length) {
+    if (uuid.length && certificationDone !== "1") {
       triggerFetchRunStatus();
     } else {
       // resetStates()
+      removeCertificationDone(); // Clear certification done status
     }
     // eslint-disable-next-line
   }, [uuid]);
```
[the remark this answers is not in the document]
Comment (amnambiar) [reply src/pages/certification/Certification.tsx]: Ignoring this suggestion

File: src/pages/certification/Certification.tsx [hunk 3]
```diff
@@ -115,7 +161,9 @@ const Certification = () => {
         );
         /** For mock */
         // const response = await postData.get('static/data/run')
-        dispatch(setUuid(response.data));
+        const runId = response.data.toString();
+        dispatch(setUuid(runId));
+        setCertificationUuid(runId);
       } catch (e) {
         handleErrorScenario();
         console.log(e);
```
[the remark this answers is not in the document]
Comment (amnambiar) [reply src/pages/certification/Certification.tsx]: a default error is displayed now, but will revisit that on revamp

File: src/pages/certification/Certification.tsx [hunk 4]
```diff
@@ -63,8 +56,46 @@ const Certification = () => {
...
+
+  // Set form as submitted on component load if certification is running
+  useEffect(() => {
+    if (startTime && endTime && certificationDone !== "1") {
+      setFormSubmitted(true);
+      setSubmitting(true);
+      form.setValue("commit", uuid);
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [uuid]);
```
[the remark this answers is not in the document]
Comment (amnambiar) [reply src/pages/certification/Certification.tsx]: ignoring this comment

File: src/pages/certification/Certification.tsx [hunk 5]
```diff
@@ -63,8 +56,46 @@ const Certification = () => {
...
+    if (certificationUuid && certificationRunTime) { 
+      const { startTime, endTime, runState } = certificationRunTime;
+      dispatch(setUuid(certificationUuid));
+      dispatch(setStates({ startTime, endTime, runState }));
+      dispatch(setBuildInfo());
+    } else {
+      resetRunForm()
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, []);
```
[the remark this answers is not in the document]
Comment (amnambiar) [reply src/pages/certification/Certification.tsx]: ignoring this comment

File: src/pages/certification/Certification.tsx [hunk 6]
```diff
@@ -11,30 +11,21 @@ import { useLogs } from "hooks/useLogs";
...
+import { clearStates, setBuildInfo, setStates, } from "./slices/logRunTime.slice";
 import { deleteTestHistoryData } from "pages/testHistory/slices/deleteTestHistory.slice";
 import { useConfirm } from "material-ui-confirm";
-import { Link } from "react-router-dom";
+import { Link, useNavigate } from "react-router-dom";
 import Loader from "components/Loader/Loader";
+import { LocalStorageKeys } from "constants/constants";
+import useLocalStorage from "hooks/useLocalStorage";
 
 const TIMEOFFSET = 1000;
```
[the remark this answers is not in the document]
Comment (amnambiar) [reply src/pages/certification/Certification.tsx:32]: It is used in 2nd argument of `useDelayedApi` hook call - `refetchMin * TIMEOFFSET`

File: src/pages/certification/certification-result/CertificationResult.tsx [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+    param.uuid as string,
+    true,
+    false,
+    handleErrorScenario
+  );
+  
+  // Show loader until data is fetched
+  if (!resultData || !Object.keys(resultData).length) {
+    return <Loader />
+  }
```
[the remark this answers is not in the document]
Comment (amnambiar) [reply src/pages/certification/certification-result/CertificationResult.tsx:89]: will be considered upon revamp

File: src/pages/certification/certification-result/CertificationResult.tsx [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+    }, 5000); // hide after 5 seconds
+    setTimelineConfig(TIMELINE_CONFIG);
+  }, []);
+
+  const { logInfo } = useLogs(
+    param.uuid as string,
+    true,
+    false,
+    handleErrorScenario
+  );
```
[the remark this answers is not in the document]
Comment (amnambiar) [reply src/pages/certification/certification-result/CertificationResult.tsx:84]: ignoring this for now, as uuid is a string

File: src/pages/certification/certification-result/CertificationResult.tsx [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+  const handleErrorScenario = useCallback(() => {
+    // show an api error toast
+    setErrorToast(true);
+    setErrorToast(true);
+    const timeout = setTimeout(() => {
+      clearTimeout(timeout);
+      setErrorToast(false);
+    }, 5000); // hide after 5 seconds
+    setTimelineConfig(TIMELINE_CONFIG);
+  }, []);
```
[the remark this answers is not in the document]
Comment (amnambiar) [reply src/pages/certification/certification-result/CertificationResult.tsx]: already done

File: src/pages/testHistory/TestHistory.tsx [hunk 10]
```diff
@@ -51,6 +52,7 @@ const TestHistory = () => {
   const certificationData: IRunCertifications = {};
   const reportData: IRunReport = {};
   const timeZone = dayjs.tz.guess()
+  const navigate = useNavigate();
 
   useEffect(() => {
     fetchTableData();
```
[the remark this answers is not in the document]
Comment (amnambiar) [reply src/pages/testHistory/TestHistory.tsx:58]: ignoring

File: src/pages/certification/Certification.tsx [hunk 11]
```diff
@@ -191,10 +214,6 @@ const Certification = () => {
     setTimelineConfig(TIMELINE_CONFIG);
   },[form])
```
[the remark this answers is not in the document]
Comment (amnambiar) [reply src/pages/certification/Certification.tsx]: ignoring

**Comment by bogdan-manole** [comment]
If there is any error during the profile completion or a previous data shape, the user will remain forever with a blank page

image

**Comment by bogdan-manole** [comment]
I do have an subscription but I'm not allowed to start a TestRun
image
image
image

**Comment by bogdan-manole** [comment]
Navigating from a previous report, having a running test in action doesn't show the state of the running test but the page to start a new one. After refresh it works
test

**Comment by bogdan-manole** [comment]
1. Abort a test run
2. pressing the IOG Home icon will bring you back to the previous state of the run, which is ok, but in the commit hash input we have autocompleted the Run ID, I guess, instead of empty or the previous commit hash
image

**Comment by bogdan-manole** [comment]
At some point the server crashed, and in the localStarage the userDetails is changed to undefined (I guess is related to the previous comment I added), now we can't load the page anymore, we have a blank page
image
image

**Comment by bogdan-manole** [comment]
1. Click view report on a previous run
2. Click on the Homepage icon
3. Start a new run test
4. The page doesn't show the run progress


https://github.com/input-output-hk/dapps-certification-web/assets/3176255/784eaf88-52b8-4952-a7d2-cd5b1568add5

File: src/components/ConnectWallet/ConnectWallet.tsx [hunk 12]
```diff
@@ -35,7 +35,7 @@ const ConnectWallet = () => {
     localStorage.getItem(LocalStorageKeys.isLoggedIn) === "true" ? true : false
   );
 
-  const [, setUserDetails] = useLocalStorage(
+  const [, setUserDetails, removeUserDetails] = useLocalStorage(
     LocalStorageKeys.userDetails,
     localStorage.getItem(LocalStorageKeys.userDetails)
       ? JSON.parse(localStorage.getItem(LocalStorageKeys.userDetails)!)
```
[the remark this answers is not in the document]
Comment (amnambiar) [reply src/components/ConnectWallet/ConnectWallet.tsx:66]: Will refine this hook altogether later

File: src/pages/userProfile/UserProfile.tsx [hunk 13]
```diff
@@ -161,7 +161,7 @@ const UserProfile = () => {
         }
         setShowError(errorMsg);
         const timeout = setTimeout(() => { clearTimeout(timeout); setShowError("") }, 5000)
-        setUserDetails({ dapp: null });        
+        removeUserDetails();        
       })
     };
     submitProfile();
```
[the remark this answers is not in the document]
Comment (amnambiar) [reply src/pages/userProfile/UserProfile.tsx:167]: TBD

File: src/pages/userProfile/UserProfile.tsx [hunk 14]
```diff
@@ -43,7 +43,7 @@ const UserProfile = () => {
   const [searchParams, setSearchParams] = useSearchParams();
   const githubAccessCode = searchParams.get("code");
 
-  const [, setUserDetails] = useLocalStorage(
+  const [, setUserDetails, removeUserDetails] = useLocalStorage(
     LocalStorageKeys.userDetails,
     localStorage.getItem(LocalStorageKeys.userDetails)
       ? JSON.parse(localStorage.getItem(LocalStorageKeys.userDetails)!)
```
[the remark this answers is not in the document]
Comment (amnambiar) [reply src/pages/userProfile/UserProfile.tsx:49]: Will refine this hook altogether later

File: src/pages/certification/certification-result/CertificationResult.tsx [hunk 15]
```diff
@@ -0,0 +1,146 @@
...
+          const resultJson = isArrayResult
+            ? res.data.result[0]
+            : res.data.result;
+          if (isArrayResult) {
+            setCoverageFile(res.data.result[1]);
+          }
+          const unitTestResult = processFinishedJson(resultJson);
+          setUnitTestSuccess(unitTestResult);
+          setResultData(resultJson);
+        }
```
[the remark this answers is not in the document]
Comment (amnambiar) [reply src/pages/certification/certification-result/CertificationResult.tsx]: If not finished, no states have to be altered and just progress the timeline. Hence this

File: src/pages/certification/Certification.tsx [hunk 16]
```diff
@@ -63,16 +55,46 @@ const Certification = () => {
...
+    removeCommitHash()
+    dispatch(clearUuid())
+    dispatch(clearStates())
+    setSubmitting(false)
+    setFormSubmitted(false)
+    form.reset()
+  }
 
   const resetStates = () => {
     setRunState("")
```
[the remark this answers is not in the document]
Comment (amnambiar) [reply src/pages/certification/Certification.tsx]: Will refine the useLocalStorage later. fixed setApiFetching

File: src/pages/certification/certification-result/CertificationResult.tsx [hunk 17]
```diff
@@ -0,0 +1,146 @@
...
```
[the remark this answers is not in the document]
Comment (amnambiar) [reply src/pages/certification/certification-result/CertificationResult.tsx:56]: TBD

File: src/pages/certification/certification-result/CertificationResult.tsx [hunk 18]
```diff
@@ -0,0 +1,146 @@
...
+    param.uuid as string,
+    true,
+    false,
+    handleErrorScenario
+  );
+  
+  // Show loader until data is fetched
+  if (!resultData || !Object.keys(resultData).length) {
+    return <Loader />
+  }
```
[the remark this answers is not in the document]
Comment (amnambiar) [reply src/pages/certification/certification-result/CertificationResult.tsx:89]: TBD

File: src/pages/certification/Certification.tsx [hunk 19]
```diff
@@ -63,8 +56,46 @@ const Certification = () => {
...
+  // Set form as submitted on component load if certification is running
+  useEffect(() => {
+    if (startTime && endTime && certificationDone !== "1") {
+      setFormSubmitted(true);
+      setSubmitting(true);
+      form.setValue("commit", uuid);
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [uuid]);
+
```
[the remark this answers is not in the document]
Comment (amnambiar) [reply src/pages/certification/Certification.tsx]: Refining the hook is TBD

File: src/components/TimelineItem/timeline.helper.tsx [hunk 20]
```diff
@@ -46,3 +46,38 @@ export const processFinishedJson = (result: { [x: string]: any }): boolean => {
...
+        } else {
+          returnObj["progress"] = 0;
+        }
+      }
+      return returnObj;
+    }
+    // Set the previously executed states as passed
+    return setManyStatus(index, config, item, status, "passed");
+  });
+}
```
[the remark this answers is not in the document]
Comment (amnambiar) [reply src/components/TimelineItem/timeline.helper.tsx:83]: Ignoring for now

**Comment by amnambiar** [comment]
@bogdan-manole / @RSoulatIOHK ,

Fixed
* Subscription is active but unable to start test run - resolved in https://github.com/input-output-hk/dapps-certification-web/pull/6

Open
* Blank page with isLoggedIn:true and userDetails:undefined in localStorage - unable to find exact cause of issue, is existent in master as well. Raised [PLT-7439](https://input-output.atlassian.net/browse/PLT-7439)
* Bug - Click on View Report in Testing History, while in Test Result UI view hit browser back, notice you have to hit the back twice to take you back to Testing History. Will recheck the navigation in the revamped UI and fix this bug there. Raised [PLT-7425](https://input-output.atlassian.net/browse/PLT-7425)

Other state pertaining issues are resolved. Please re-review

Cc: @EehMauro

File: src/pages/certification/certification-result/CertificationResult.tsx [hunk 21]
```diff
@@ -0,0 +1,145 @@
...
+          setUnitTestSuccess(unitTestResult);
+          setResultData(resultJson);
+        }
+      } catch (error) {
+        handleErrorScenario();
+        console.error('Failed:', error);
+      }
+    })();
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [param.uuid]);
```
[the remark this answers is not in the document]
Comment (amnambiar) [reply src/pages/certification/certification-result/CertificationResult.tsx:63]: ignoring

File: src/pages/certification/Certification.tsx [hunk 22]
```diff
@@ -115,10 +161,12 @@ const Certification = () => {
...
+        const runId = response.data.toString();
+        dispatch(setUuid(runId));
+        setCertificationUuid(runId);
       } catch (e) {
         handleErrorScenario();
-        console.log(e);
+        console.error('Failed:', e);
       }
     };
     triggerAPI();
```
[the remark this answers is not in the document]
Comment (amnambiar) [reply src/pages/certification/Certification.tsx]: ignoring for now

File: src/pages/certification/Certification.tsx [hunk 23]
```diff
@@ -307,6 +322,8 @@ const Certification = () => {
       {formSubmitted && (
         <>
           <div id="resultContainer" data-testid="resultContainer">
+            {/* 
+            **TBD -- migrated into CertificationResult**
             {runStatus === "finished" ? (
               <button
                 className="back-btn"
```
[the remark this answers is not in the document]
Comment (amnambiar) [reply src/pages/certification/Certification.tsx]: ignoring for now

File: src/pages/certification/certification-result/CertificationResult.tsx [hunk 24]
```diff
@@ -0,0 +1,146 @@
...
+            setCoverageFile(res.data.result[1]);
+          }
+          const unitTestResult = processFinishedJson(resultJson);
+          setUnitTestSuccess(unitTestResult);
+          setResultData(resultJson);
+        }
+      } catch (error) {
+        handleErrorScenario();
+        console.error('Failed:', error);
+      }
```
[the remark this answers is not in the document]
Comment (amnambiar) [reply src/pages/certification/certification-result/CertificationResult.tsx:60]: ignoring for now

File: src/pages/certification/certification-result/CertificationResult.tsx [hunk 25]
```diff
@@ -0,0 +1,146 @@
...
+              <>
+                <Button
+                  className="report-download"
+                  onClick={(_) => handleDownloadResultData(resultData)}
+                  buttonLabel="Download Report"
+                  iconUrl={DownloadIcon}
+                />
+                {state?.certifiable ? <CreateCertificate uuid={param.uuid as string} /> : null}
+              </>
+            ) : null}
```
[the remark this answers is not in the document]
Comment (amnambiar) [reply src/pages/certification/certification-result/CertificationResult.tsx:107]: ignoring for now, for revamp

File: src/pages/certification/certification-result/CertificationResult.tsx [hunk 26]
```diff
@@ -0,0 +1,146 @@
...
+          }
+          const unitTestResult = processFinishedJson(resultJson);
+          setUnitTestSuccess(unitTestResult);
+          setResultData(resultJson);
+        }
+      } catch (error) {
+        handleErrorScenario();
+        console.error('Failed:', error);
+      }
+    })();
```
[the remark this answers is not in the document]
Comment (amnambiar) [reply src/pages/certification/certification-result/CertificationResult.tsx]: ignoring for now, for revamp

File: src/pages/certification/certification-result/CertificationResult.tsx [hunk 27]
```diff
@@ -0,0 +1,145 @@
...
+
+  const handleErrorScenario = useCallback(() => {
+    // show an api error toast
+    setErrorToast(true);
+    const timeout = setTimeout(() => {
+      clearTimeout(timeout);
+      setErrorToast(false);
+    }, 5000); // hide after 5 seconds
+    setTimelineConfig(TIMELINE_CONFIG);
+  }, []);
```
[the remark this answers is not in the document]
Comment (amnambiar) [reply src/pages/certification/certification-result/CertificationResult.tsx:77]: ignoring for now

File: src/pages/certification/certification-result/CertificationResult.tsx [hunk 28]
```diff
@@ -0,0 +1,145 @@
...
+    param.uuid as string,
+    true,
+    false,
+    handleErrorScenario
+  );
+  
+  // Show loader until data is fetched
+  if (!resultData || !Object.keys(resultData).length) {
+    return <Loader />
+  }
```
[the remark this answers is not in the document]
Comment (amnambiar) [reply src/pages/certification/certification-result/CertificationResult.tsx:89]: ignoring for now, for revamp

File: src/pages/certification/certification-result/CertificationResult.tsx [hunk 29]
```diff
@@ -0,0 +1,145 @@
...
+            {Object.keys(resultData).length ? (
+              <>
+                <Button
+                  className="report-download"
+                  onClick={(_) => handleDownloadResultData(resultData)}
+                  buttonLabel="Download Report"
+                  iconUrl={DownloadIcon}
+                />
+                {state?.certifiable ? <CreateCertificate uuid={param.uuid as string} /> : null}
+              </>
```
[the remark this answers is not in the document]
Comment (amnambiar) [reply src/pages/certification/certification-result/CertificationResult.tsx:106]: ignoring for now

File: src/components/AvatarDropdown/AvatarDropdown.tsx [hunk 30]
```diff
@@ -6,26 +6,15 @@ import MenuItem from "@mui/material/MenuItem";
 import Avatar from "@mui/material/Avatar";
 import "./AvatarDropDown.scss";
 import DropoverMenu from "components/DropoverMenu/DropoverMenu";
-import useLocalStorage from "hooks/useLocalStorage";
-import { LocalStorageKeys } from 'constants/constants';
+// import useLocalStorage from "hooks/useLocalStorage";
```
Comment (bogdan-manole) [inline src/components/AvatarDropdown/AvatarDropdown.tsx]: can we remove the commented code?

**Review by bogdan-manole — CHANGES_REQUESTED** [review]
just a small reminder to delete commented code

**Review by bogdan-manole — APPROVED** [review]

**Pull request merged**